static/app/components/searchSyntax/renderer.tsx

```diff
@@ -149,6 +156,78 @@ function FilterToken({
   );
 }
 
+function FreeTextToken({
+  filter,
+  cursor,
+}: {
+  cursor: number;
+  filter: TokenResult<Token.FreeText>;
+}) {
+  const isActive = isWithinToken(filter, cursor);
+
+  // This state tracks if the cursor has left the filter token. We initialize it
+  // to !isActive in the case where the filter token is rendered without the
+  // cursor initially being in it.
+  const [hasLeft, setHasLeft] = useState(!isActive);
+
+  // Used to trigger the shake animation when the element becomes invalid
+  const filterElementRef = useRef<HTMLSpanElement>(null);
+
+  // Trigger the effect when isActive changes to updated whether the cursor has
+  // left the token.
+  useEffect(() => {
+    if (!isActive && !hasLeft) {
+      setHasLeft(true);
+    }
+  }, [hasLeft, isActive]);
+
+  const showInvalid = hasLeft && !!filter.invalid;
+  const showTooltip = showInvalid && isActive;
+
+  const reduceMotion = useReducedMotion();
+
+  // Trigger the shakeAnimation when showInvalid is set to true. We reset the
+  // animation by clearing the style, set it to running, and re-applying the
+  // animation
+  useEffect(() => {
+    if (!filterElementRef.current || !showInvalid || reduceMotion) {
+      return;
+    }
+
+    const style = filterElementRef.current.style;
+
+    style.animation = 'none';
+    void filterElementRef.current.offsetTop;
+
+    window.requestAnimationFrame(
+      () => (style.animation = `${shakeAnimation.name} 300ms`)
+    );
+  }, [reduceMotion, showInvalid]);
+
+  if (!filter.invalid?.reason) {
+    return <Fragment>{renderToken(filter.value, cursor)}</Fragment>;
+  }
```

Comment:
A lot of this is duplicated from `filterToken` right?

Can we actually just turn this into a hook in this file that we reuse, maybe named like `useTokenPresentationState`

Reply:
I can work on improving this in a separate PR if that is ok 🙂

Reply:
Sounds good, let's definitely follow up right after though! right now it's just straight duplicated I think haha

let's at least get the variable names right in this iteration https://github.com/getsentry/sentry/pull/49604/files#r1205903613